packages/eslint-plugin/src/rules/unbound-method.ts

```diff
@@ -209,6 +214,42 @@ export default createRule<Options, MessageIds>({
               );
             }
           });
+          return;
+        }
+
+        if (node.typeAnnotation) {
+          const subtypes = tsutils.unionTypeParts(
+            services.getTypeAtLocation(node.typeAnnotation.typeAnnotation),
+          );
+          for (const property of node.properties) {
+            for (const subtype of subtypes) {
+              if (
+                property.type !== AST_NODE_TYPES.Property ||
+                property.key.type !== AST_NODE_TYPES.Identifier
+              ) {
+                continue;
+              }
+
+              const propertyKey = property.key;
+              if (
+                checkIfMethodAndReport(
+                  propertyKey,
+                  subtype.getProperty(propertyKey.name),
+                ) ||
+                (tsutils.isIntersectionType(subtype) &&
+                  tsutils
+                    .intersectionTypeParts(subtype)
+                    .some(intersectionSubtype =>
+                      checkIfMethodAndReport(
```

Comment:
Array callbacks should be pure functions. Please use explicit looping for control flow that includes `checkIfMethodAndReport`.